Sources/CADCacheAdvance/CADCacheAdvance.swift

```diff
@@ -0,0 +1,108 @@
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+#if SWIFT_PACKAGE_MANAGER
+// Swift Package Manager defines multiple modules, while other distribution mechanisms do not.
+// We only need to import CacheAdvance if this project is being built with Swift Package Manager.
+import CacheAdvance
+#endif
+import Foundation
+
+/// A cache that enables the performant persistence of individual messages to disk.
+/// This cache is intended to be written to and read from using the same serial queue.
+/// - Attention: This type is meant to be used by Objective-C code, and is not exposed to Swift. Swift code should use CacheAdvance<T>.
+@objc(CADCacheAdvance)
+@available(swift, obsoleted: 1.0)
+public final class __ObjectiveCCompatibleCacheAdvanceWithGenericStorage: NSObject {
+
+    // MARK: Initialization
+
+    /// Creates a new instance of the receiver.
+    ///
+    /// - Parameters:
+    ///   - fileURL: The file URL indicating the desired location of the on-disk store. This file should already exist.
+    ///   - maximumBytes: The maximum size of the cache, in bytes. Logs larger than this size will fail to append to the store.
+    ///   - shouldOverwriteOldMessages: When `true`, once the on-disk store exceeds maximumBytes, new entries will replace the oldest entry.
+    ///
+    /// - Warning: `maximumBytes` must be consistent for the life of a cache. Changing this value after logs have been persisted to a cache will prevent appending new messages to this cache.
+    /// - Warning: `shouldOverwriteOldMessages` must be consistent for the life of a cache. Changing this value after logs have been persisted to a cache will prevent appending new messages to this cache.
+    @objc
+    public init(
+        fileURL: URL,
+        maximumBytes: Bytes,
+        shouldOverwriteOldMessages: Bool)
+        throws
+    {
+        cache = try CacheAdvance<Storage>(
+            fileURL: fileURL,
+            maximumBytes: maximumBytes,
+            shouldOverwriteOldMessages: shouldOverwriteOldMessages)
+    }
+
+    // MARK: Public
+
+    @objc
+    public var fileURL: URL {
+        cache.fileURL
+    }
+
+    /// Checks if the all the header metadata provided at initialization matches the persisted header. If not, the cache is not writable.
+    /// - Returns: `true` if the cache is writable.
+    @objc
+    public var isWritable: Bool {
+        (try? cache.isWritable()) ?? false
+    }
+
+    /// Appends a message to the cache.
+    /// - Parameter message: A message to write to disk. Must be smaller than both `maximumBytes - FileHeader.expectedEndOfHeaderInFile` and `MessageSpan.max`.
+    @objc
+    public func appendMessage(_ message: Data) throws {
```

Comment:
This is a bit off topic from this PR, but I'm going to add the comment anyway since this PR made me think of it.
Do we expect this data to be String? Or can it truly be any raw stream of bytes?
I guess I'm curious how non-string data gets encoded to JSON.
It seems to work but I'm curious how it's represented...

Reply:
(Though I guess I'm not sure we have any tests for non-UTF-8 data? It could be interesting to add one just to be sure, if we haven't explicitly tested that use case anywhere.)

Reply:
This `message` could be anything. It's likely a string, but it could easily be any serialized type. Once we have an object as Data, we treat it as Data (we do not assume anything about it, except that it can be put in a Codable object and then serialized as JSON).

Given my performance measurement work on the other branch, I'm starting to wonder if I need to take a different approach to this PR.

The thin Objective-C wrapper on top of CacheAdvance is great, in that it minimizes the maintenance burden for me. However, it does mean that messages sent to `CADCacheAdvance` are likely being encoded twice. Once `Data` so that it can be sent to `CADCacheAdvance`, and then again as `Storage` so that `CacheAdvance` can persist them.

I may need to offer a way to customize serialization/deserialization code within `CacheAdvance` (at least internally probably not worth making that API public). Going to sit on this thought for a bit.

Reply:
Just pushed up performance tests (which required duplicating the lorum ipsum array 😞). `test_performance_append_fillableCache` for `CADCacheAdvance` is 0.06s slower (0.017s) than the same test for `CacheAdvance` (0.011s).

That feels worth fixing.

Reply:
Sigh... to customize this behavior I will need to make the new API public, because `CADCacheAdvance` lives in a separate library 😫

Reply:
Okay okay we've got 2910f9ddbce71e2f4ecba10123e9e42d4532dbf4 now. `CADCacheAdvance`'s `test_performance_append_fillableCache` is now taking 0.011s.

Reply:
One nicety about the new public API is that if someone has a more performant way to serialize their messages (i.e. using something other than JSON encoding/decoding), they can now use it. Feels reasonable to let our consumers eek out as much performance from our system as they are able.